### PR TITLE
NIFI-8030: Added property to ReportLineageToAtlas to configure file/d…

### PR DESCRIPTION
**Pull request by turcsanyip**
…irectory level logging of [hd]fs_path entities to Atlas

https://issues.apache.org/jira/browse/NIFI-8030

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/FilesystemPathsLevel.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.atlas.provenance;
+
+public enum FilesystemPathsLevel {
+    File, Directory
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/FilesystemPathsLevel.java:20]: Could you please rename the constants to be uppercase?

File: nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/analyzer/FilePath.java [hunk 2]
```diff
@@ -56,11 +59,18 @@ public DataSetRefs analyze(AnalysisContext context, ProvenanceEventRecord event)
...
+        final String path;
+        if (context.getFilesystemPathsLevel() == FilesystemPathsLevel.Directory) {
+            final String dirPath = StringUtils.substringBeforeLast(uri.getPath(), PATH_SEPARATOR);
+            path = dirPath.isEmpty() ? PATH_SEPARATOR : dirPath;
+        } else {
+            path = uri.getPath();
+        }
+
         ref.set(ATTR_NAME, path);
         ref.set(ATTR_PATH, path);
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/analyzer/FilePath.java:75]: What do you think of extracting these lines to a separate method? The scode fragment is duplicated in HDFSParth:52-61.

File: nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java [hunk 3]
```diff
@@ -37,13 +37,15 @@
     private final NamespaceResolver namespaceResolver;
     private final ProvenanceRepository provenanceRepository;
     private final String awsS3ModelVersion;
+    private final FilesystemPathsLevel filesystemPathsLevel;
 
     public StandardAnalysisContext(NiFiFlow nifiFlow, NamespaceResolver namespaceResolver,
-                                   ProvenanceRepository provenanceRepository, String awsS3ModelVersion) {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java:42]: I think the arguments could be final.

File: nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/FilesystemPathsLevel.java (hunk 1 above)
Comment (turcsanyip) [reply nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/FilesystemPathsLevel.java]: done

File: nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/analyzer/FilePath.java (hunk 2 above)
Comment (turcsanyip) [reply nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/analyzer/FilePath.java]: done

File: nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java (hunk 3 above)
Comment (turcsanyip) [reply nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java:42]: In general, I do not prefer using `final` for local variables and parameters. All these variables should be final in most cases and IDEs can mark it automatically if a variable is used in a non final way.
For existing codes I follow the existing pattern in the class/method and only add `final` if it is already used there.
This parameter list does not use `final`-s so I would not add it for the new param either.

**Review by pvillard31 — APPROVED** [review]

**Comment by pvillard31** [comment]
+1, merged to main, thanks @turcsanyip and @Lehel44 !

**Pull request closed without merging**